tests/functional/base.py

```diff
@@ -110,19 +180,145 @@ def custodia_server(self, simple_configuration, request, dev_null):
             stdout=stdout, stderr=stderr
         )
 
-        self._wait_pid(self.process, 2)
-        self._wait_socket(self.process, 5)
+        wait_pid(self.process, 2)
+        wait_socket(self.process, self.env['CUSTODIA_SOCKET'], 5)
 
         arg = '{}/custodia.sock'.format(CustodiaServerRunner.test_dir)
         url = 'http+unix://{}'.format(url_escape(arg, ''))
         self.custodia_client = CustodiaHTTPClient(url)
 
         def fin():
             self.process.terminate()
-            if not self._wait_pid(self.process, 2):
+            if not wait_pid(self.process, 2):
                 self.process.kill()
-                if not self._wait_pid(self.process, 2):
+                if not wait_pid(self.process, 2):
                     raise AssertionError("Hard kill failed")
 
         request.addfinalizer(fin)
         return self.custodia_client
+
+
+@pytest.mark.servertest
+class CustodiaTestEnvironment(UniqueNumber):
+    test_dir = 'tests/functional/tmp_auth_plugin'
+
+    @classmethod
+    def setup_class(cls):
+        if os.path.isdir(cls.test_dir):
+            shutil.rmtree(cls.test_dir)
+        os.makedirs(cls.test_dir)
+
+    @classmethod
+    def teardown_class(cls):
+        shutil.rmtree(cls.test_dir)
+
+    def reset_environment(self):
+        if os.path.isdir(self.test_dir):
+            shutil.rmtree(self.test_dir)
+        os.makedirs(self.test_dir)
+
+
+class AuthPlugin(Enum):
+    SimpleCredsAuth = 1
+    SimpleHeaderAuth = 2
+    SimpleAuthKeys = 3
+    SimpleClientCert = 4
+
+
+class CustodiaServer(object):
+    def __init__(self, test_dir, conf_params):
+        self.process = None
+        self.custodia_client = None
+        self.test_dir = test_dir
+        self.custodia_conf = os.path.join(self.test_dir, 'custodia.conf')
+        self.params = conf_params
+
+        self.out_fd = os.open(os.devnull, os.O_RDWR)
```

Comment:
This leaks a fd when __enter__ fails. My original code used a global scoped fixture to have a single FD for all tests.